web/pdf_find_controller.js

```diff
@@ -126,12 +126,35 @@ function normalize(text) {
   } else {
     // Compile the regular expression for text normalization once.
     const replace = Object.keys(CHARACTERS_TO_NORMALIZE).join("");
-    const toNormalizeWithNFKC =
-      "\u2460-\u2473" + // Circled numbers.
-      "\u24b6-\u24ff" + // Circled letters/numbers.
-      "\u3244-\u32bf" + // Circled ideograms/numbers.
-      "\u32d0-\u32fe" + // Circled ideograms.
-      "\uff00-\uffef"; // Halfwidth, fullwidth forms.
+    const ranges = [];
+    const range = [];
+    const diacriticsRegex = /^\p{M}$/u;
+    // Some chars must be replaced by their NFKC counterpart during a search.
+    for (let i = 0; i < 65536; i++) {
+      const c = String.fromCharCode(i);
+      if (c.normalize("NFKC") !== c && !diacriticsRegex.test(c)) {
+        if (range.length !== 2) {
+          range[0] = range[1] = i;
+          continue;
+        }
+        if (range[1] + 1 !== i) {
+          if (range[0] === range[1]) {
+            ranges.push(c);
+          } else {
+            ranges.push(
+              `${String.fromCharCode(range[0])}-${String.fromCharCode(
+                range[1]
+              )}`
+            );
+          }
+          range[0] = range[1] = i;
+        } else {
+          range[1] = i;
+        }
+      }
+    }
```

Comment:
How does all of this "extra" parsing here affect performance of the initial search operation?

Reply:
On my computer it takes around 7,8 ms (I added some `performance.mark` around that code and then measure with the profiler).
The final string has a length of 315 chars.
We can hard-code the string but if there are some changes in normalization stuff (on the js engine side) we would have to regenerate that string, but maybe it's something which will never happen, I don't know.
So if you think it's too slow, then we hard-code the string and we wait for complaints.

Reply:
> On my computer it takes around 7,8 ms (I added some `performance.mark` around that code and then measure with the profiler).

That doesn't look too bad on its own, although I guess this is on powerful hardware.
However possibly a better question, on my part, would've been how much time *in percentage* of the initial `normalize`-call is spent in this new code-block?

> We can hard-code the string but if there are some changes in normalization stuff (on the js engine side) we would have to regenerate that string, but maybe it's something which will never happen, I don't know.

That's indeed a risk if we just hard-code this data, although I'd be surprised if this changes a lot.

> So if you think it's too slow, then we hard-code the string and we wait for complaints.

How about a compromise? What if we hard-code the data, but let the existing find unit-tests check that normalization hasn't changed by instead moving this logic into a helper function in `pdf_find_utils.js`?
```diff
diff --git a/web/pdf_find_utils.js b/web/pdf_find_utils.js
index 24ec4c575..60964973b 100644
--- a/web/pdf_find_utils.js
+++ b/web/pdf_find_utils.js
@@ -112,4 +112,47 @@ function getCharacterType(charCode) {
   return CharacterType.ALPHA_LETTER;
 }
 
-export { CharacterType, getCharacterType };
+let NormalizeWithNFKC;
+
+function getNormalizeWithNFKC() {
+  NormalizeWithNFKC ||=
+    "¨ª¯²²-µ¸-º¼-¾Ĳ-ĳĿ-ŀſǄǄ-ǌǱ-ǳʰ-ʸ˘-˝ˠ-ˤͺ;΄΄-΅ϐϐ-ϖϰ-ϲϴ-ϵևٵٵ-ٸक़-य़ড়-ঢ়ਲ਼ਸ਼ਖ਼ਖ਼-ਜ਼ଡ଼ଡ଼-ଢ଼ຳໜໜ-ໝགྷཌྷདྷབྷཛྷཀྵჼᴬᴬ-ᴮᴰ-ᴺᴼ-ᵍᵏ-ᵪᶛᶛ-ᶿẚ-ẛέήίόύώΆ᾽᾽-῁Ή῍῍-῏Ί῝῝-῟Ύ῭῭-`Ώ´´-῾ - ‗․․-…″″-‴‶-‷‾⁇⁇-⁉ ⁰⁰-ⁱ⁴-₎ₐ-ₜ℀℀-℃℅-ℇ℉-ℓℕ-№ℙ-ℝ℠-™ΩℨKK-ℭℯ-ℱℳ-ℹ℻-⅀ⅅ-ⅉ⅐-ⅿ∬∬-∭∯-∰〈-〉①-⓪⩴⩴-⩶ⱼⱼ-ⱽ⺟⻳⼀⼀-⿕〶〸〸-〺゛-゜ヿㄱㄱ-ㆎ㆒-㆟㈀-㈞㈠-㉇㉐-㉾㊀-㏿ꚜ-ꚝꟲꟲ-ꟴꟸ-ꟹꭜ-ꭟ豈豈-嗀晴凞凞-羽諸逸逸-都飯-舘並-龎ﬀ-ﬆﬓ-ﬗײַײַ-זּטּ-לּנּנּ-סּףּ-פּצּ-ﮱﯓ-ﴽﵐ-ﶏﶒ-ﷇﷰ-﷼︐-︙︰-﹄﹇-﹒﹔-﹦﹨-﹫ﹰ-ﹲﹶﹶ-ﻼ！-ﾾￂ-ￇￊ-ￏￒ-ￗￚ-ￜ￠-￦";
+
+  if (typeof PDFJSDev === "undefined" || PDFJSDev.test("TESTING")) {
+    const ranges = [];
+    const range = [];
+    const diacriticsRegex = /^\p{M}$/u;
+    // Some chars must be replaced by their NFKC counterpart during a search.
+    for (let i = 0; i < 65536; i++) {
+      const c = String.fromCharCode(i);
+      if (c.normalize("NFKC") !== c && !diacriticsRegex.test(c)) {
+        if (range.length !== 2) {
+          range[0] = range[1] = i;
+          continue;
+        }
+        if (range[1] + 1 !== i) {
+          if (range[0] === range[1]) {
+            ranges.push(c);
+          } else {
+            ranges.push(
+              `${String.fromCharCode(range[0])}-${String.fromCharCode(
+                range[1]
+              )}`
+            );
+          }
+          range[0] = range[1] = i;
+        } else {
+          range[1] = i;
+        }
+      }
+    }
+    if (ranges.join("") !== NormalizeWithNFKC) {
+      throw new Error(
+        "getNormalizeWithNFKC - update the `NormalizeWithNFKC` string."
+      );
+    }
+  }
+  return NormalizeWithNFKC;
+}
+
+export { CharacterType, getCharacterType, getNormalizeWithNFKC };
```